databuilder/dsl.py

```diff
@@ -297,21 +279,82 @@ def __invert__(self) -> PatientSeries:
         return PatientSeries(value=comparator_value)
 
     def __repr__(self) -> str:
-        return f"PatientSeries(value={self.value})"
+        return f"{self.__class__.__name__}(value={self.value})"
 
-    def __hash__(self) -> int:
-        return hash(repr(self.value))
 
-    def round_to_first_of_month(self):
-        return PatientSeries(RoundToFirstOfMonth(self.value))
+class BoolSeries(PatientSeries):
+    def __and__(self, other: BoolSeries) -> BoolSeries:
+        return BoolSeries(value=self.value & other.value)
+
+    def __or__(self, other: BoolSeries) -> BoolSeries:
+        return BoolSeries(value=self.value | other.value)
+
+
+class CodeSeries(PatientSeries):
+    def __and__(self, other: CodeSeries) -> BoolSeries:
+        return BoolSeries(value=self.value & other.value)
+
+    def __invert__(self) -> BoolSeries:
+        return BoolSeries(
+            value=(
+                Comparator(lhs=self.value, operator="__ne__", rhs=None, negated=True)
+            )
+        )
+
+
+class DateSeries(PatientSeries):
+    def __gt__(self, other: ValueExpression) -> BoolSeries:
+        return BoolSeries(value=self.value > other)
+
+    def __ge__(self, other: ValueExpression) -> BoolSeries:
+        return BoolSeries(value=self.value >= other)
 
-    def round_to_first_of_year(self):
-        return PatientSeries(RoundToFirstOfYear(self.value))
+    def __lt__(self, other: ValueExpression) -> BoolSeries:
+        return BoolSeries(value=self.value < other)
```

Comment:
Should this (and similar) be `BoolSeries(value=self.value < other.value)`?

Reply:
No!  `ValueExpression` is defined as `Union[PatientSeries, Comparator, str, int]`, so it's valid to pass in say `"2020-01-01"` (`test_categorise` does this).  Discovered because I made exactly this change and found tests broke… This one caught me out though, so maybe the type of `other` should be more specific for the Series types?

Reply:
I don't entirely trust `ValueExpression`. I think that for each method like this in the DSL we should carefully consider what types should be valid and specify them explicitly.

Reply:
Ah, I understand the `"2020-01-01"` case now.

But I think this might be wrong for other cases.

For instance, in this test...

```

def test_intseries_lt(int_series):
    series1 = int_series()
    series2 = int_series()

    expected = Comparator(
        operator="__lt__",
        lhs=series1.value,
        rhs=series2,
    )

    output = series1 < series2
    assert repr(output) == repr(BoolSeries(value=expected))
```

...I'm surprised that `lhs` and `rhs` in the comparator are different kinds of things.  And specifically, `rhs` is a DSL object, and the QM shouldn't know about the DSL.

I think this shows the limitations of this kind of testing.  So I think before this branch lands, we should have at least one integration test showing that a dataset definition that uses these kinds of boolean operations actually work to generate the expected data.

Am aware that this is a massive extra chunk of work!  I'd suggest parking this PR for now, and starting to work on integrating the new DSL into the integration tests.  I'd also suggest discussing approaches in Slack, as @rebkwok or @evansd might have ideas about how to go about doing this.

Reply:
I actually might have some of those sorts of tests in the date arithmetic branch.  `test_query_engine.py` is a good place to find simple tests that check generated data - they all use the old DSL, so rewriting (some of) them with the new DSL and ensuring they generate the same data would be useful.

Comparators should be expecting either a Value or a Comparator as the lhs, and a Value, Comparator or simple int/str etc as rhs.  I don't think they should ever get a DSL series (as either lhs or rhs).

Reply:
I've tracked this down to the Series comparison methods naively using `other` instead of `other.value`.  This breaks for primitive types so more to fix here.

> Comparators should be expecting either a Value or a Comparator as the lhs

@rebkwok – this is really useful to have spelled out, it hadn't really occurred to be before, thank you!

> So I think before this branch lands, we should have at least one integration test showing that a dataset definition that uses these kinds of boolean operations actually work to generate the expected data.

@inglesp – done in #314, and found some bugs, good call 👍